### DISCUSSION TITLE
Antlr dependencies and versions

### DISCUSSION THREAD

**Discussion by merks** [eclipse-xtext/xtext#2794] [Q&A] [answered]
I'm wondering about the rather strict version range dependencies on antrl 3.x:

![image](https://github.com/eclipse/xtext/assets/208716/48423394-fa55-489f-a916-14ff3ba79264)

Probably there is a good reason for that strictness. 

There are many newer versions:

![image](https://github.com/eclipse/xtext/assets/208716/e9560da4-0f33-4d40-b630-7e22e060457d)

All of these require a BND recipe because they are not available as OSGi bundles.

So to migrate away from EBR, I would expect to provide the 3.5.3 version, but that doesn't look to be compatible with the ranges I see...

The antrl4 support is available as OSGi bundles:

![image](https://github.com/eclipse/xtext/assets/208716/f2c8921d-e030-4b2e-a373-d40e29fd42fc)

But the package names and bundle names are different, so I guess there are no plans to ever use that?

**Comment by szarnekow** [comment, marked as answer]
The Xtext parser uses internal state of the Antlr parser to drive content assist proposals. Even tiny changes to the way that state is being maintained or tiny changes in the code generation templates that Antlr uses, might cause severe regressions. That's why the dependency is pinned to the seemingly old version 3.2.0.

Antlr 4 uses a complete different parsing mechanism which does essentially drop the idea of static ambiguity analysis. That's something we didn't want to loose. Performance experiments based on rather complex grammars did also show some regressions compared to Antlr 3. So a migration from 3 to 4 wasn't a trivial exercise and there was no clear benefit from doing so.

**Comment by merks** [comment]
I made a special case to pin this older version:

https://github.com/eclipse-orbit/orbit-simrel/blob/086b9cca108497635ad04ae7a0313ff12ff9d3c9/maven-bnd/tp/MavenBND.target#L1737-L1778

But I won't do that for any other version in the future. 

https://github.com/eclipse-orbit/orbit-simrel/blob/086b9cca108497635ad04ae7a0313ff12ff9d3c9/maven-bnd/tp/MavenBND.target#L1781-L1822
